eng/common/TestResources/New-TestResources.ps1

```diff
@@ -426,6 +382,82 @@ try {
         }
     }
 
+    # If no test application ID was specified during an interactive session, create a new service principal.
+    if (!$CI -and !$TestApplicationId) {
+        # Cache the created service principal in this session for frequent reuse.
+        $servicePrincipal = if ($AzureTestPrincipal -and (Get-AzADServicePrincipal -ApplicationId $AzureTestPrincipal.ApplicationId) -and $AzureTestSubscription -eq $SubscriptionId) {
+            Log "TestApplicationId was not specified; loading cached service principal '$($AzureTestPrincipal.ApplicationId)'"
+            $AzureTestPrincipal
+        } else {
+            Log "TestApplicationId was not specified; creating a new service principal in subscription '$SubscriptionId'"
+            $suffix = (New-Guid).ToString('n').Substring(0, 4)
+
+            # Service principals in the Microsoft AAD tenant must end with an @microsoft.com domain; those in other tenants
+            # are not permitted to do so. Format the non-MS name so there's an assocation with the Azure SDK.
+            if ($TenantId -eq '72f988bf-86f1-41af-91ab-2d7cd011db47') {
+                $displayName = "test-resources-$($baseName)$suffix.microsoft.com"
+            } else {
+                $displayName = "$($baseName)$suffix.test-resources.azure.sdk"
+            }
+
+            $servicePrincipal = Retry {
+                New-AzADServicePrincipal -Role "Owner" -Scope "/subscriptions/$SubscriptionId/resourceGroups/$ResourceGroupName" -DisplayName $displayName
+            }
+
+            $global:AzureTestPrincipal = $servicePrincipal
+            $global:AzureTestSubscription = $SubscriptionId
+
+            Log "Created service principal '$($AzureTestPrincipal.ApplicationId)'"
+            $AzureTestPrincipal
+            $resourceGroupRoleAssigned = $true
+        }
+
+        $TestApplicationId = $servicePrincipal.ApplicationId
+        $TestApplicationOid = $servicePrincipal.Id
+        $TestApplicationSecret = (ConvertFrom-SecureString $servicePrincipal.Secret -AsPlainText)
+    }
+
+    # Get test application OID from ID if not already provided. This may fail if the
+    # provisioner is a service principal without permissions to query AAD. This is a
+    # critical failure, but we should prompt with possible remediation.
+    if ($TestApplicationId -and !$TestApplicationOid) {
+        Log "Attempting to query the Object ID for the test service principal"
+
+        try {
+            $testServicePrincipal = Retry {
+                Get-AzADServicePrincipal -ApplicationId $TestApplicationId
+            }
+        }
+        catch {
+            Write-Warning "The Object ID of the test application was unable to be queried. You may want to consider passing it explicitly with the 'TestApplicationOid` parameter."
+            throw $_.Exception
+        }
+
+        if ($testServicePrincipal -and $testServicePrincipal.Id) {
+            $script:TestApplicationOid = $testServicePrincipal.Id
+        }
+    }
```

Comment:
As mentioned above, this has to be done before the provisioner is set.

Reply:
Can you help me understand why?